### PR TITLE
Figure property

### PR DESCRIPTION
**Pull request by Mrngilles**
What was done:
- Adding the figure attribute as a class property
- Deprecating the figure getters and setters
- Added a `deprecated_get_set` function

I know the deprecation helper is in a weird place, and I am more than willing to change it, but I had no idea where to put it...

Also, I plan on transforming other attributes (as much as possible) to properties, if you think it is a good idea. If so, do you want a PR per attribute or should I group them?

### CODE REVIEW

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -594,11 +594,26 @@ def set_path_effects(self, path_effects):
...
 
+    @property
+    def figure(self):
+        """:class:`~matplotlib.figure.Figure` instance the artist
+        belongs to"""
+        return self._figure
+
+    @figure.setter
+    def figure(self, fig):
+        self._figure = fig
```
Comment (tacaswell) [inline lib/matplotlib/artist.py]: If you are going to do this you should add validation that artists can not be moved between figures.

**Comment by tacaswell** [comment]
I would not go down this road right now, there is an on-going effort to change all of the Artist attributes into traitlets (https://github.com/matplotlib/matplotlib/pull/4762) which will supercede this.

Sorry for the poor communication and I hope you didn't spend too much time on this!

**Comment by Mrngilles** [comment]
Okay, if the attributes are being changed to traitlets, I guess it's good. I don't know about it too much, so I cannot foresee exactly all the gain the project will get from this switch.
Don't worry about my time on this, I learned quite a bit and understood the lib a bit better.
Anyway, if any help is needed with traitlets or anything else, always ready to help :smile:

**Comment by tacaswell** [comment]
If you want to help with traitlets coordinate with @rmorshea 

At a minimum we need help with testing, etc of that branch.  Something as simple as using the traitlets branch for you daily work would be great.   Hopefully once Ryan gets the general work flow worked out, up-converting attributes -> traitlets can be parallelized.

**Pull request closed without merging**